### DISCUSSION TITLE
Computing Coverage Table

### DISCUSSION THREAD

**Discussion by Jmaccaro** [KwanLab/Autometa#232] [Q&A] [answered]
Because I made a for loop to iterate through all my samples when running calculate_read_coverage.py, my coverage table is a table with just contigs and lengths for each sample. Can I use this downstream or is there an easy way to turn those into coverages (ie normalize or something?). Thanks so much!

**Comment by evanroyrees** [comment]
Hi @Jmaccaro, thank you for reaching out. I'm not quite sure I understand your question. What downstream results are you attempting to compute? Do you need to merge all of the coverage tables together?

**Reply by Jmaccaro** [reply]
Oh sorry I realized that was no context! So I used MEGAHIT for my assemblies so I didn't have the coverage info in the name like Metaspaces does. So I ran this loop to get a coverage table:  

for i in $(ls $PATHTOFASTQ/*_L002_R1_001.fastq)
do
	echo "Processing" $i
	Name=$(basename -s _L002_R1_001.fastq $i)
	samplename="${Name}"
	samplef=${Name}_L002_R1_001.fastq
	sampler=${Name}_L002_R2_001.fastq
	assembly=${Name}.contigs.fa
	calculate_read_coverage.py \
    	--assembly $PATHTOCONTIGS/$assembly\
    	--forward_reads $PATHTOFASTQ/$samplef \
    	--reverse_reads $PATHTOFASTQ/$sampler
	echo "done with" $i
done

For each sample I got a 6 bt2 files, 1 .bam, 1 .txt, 1 .tab. .. I also got only one file called coverage.tab but it doesn't look like it has all my samples and I actually can't tell which sample it was made from. The .tab files for each sample look like they contain the sequence/contig ID and length. They don't have coverage values. like the coverage.tab file has. However, I didn't get a coverage.tab file for each sample so I was hoping to be able to use the .tab files with sequence and length to start creating MAGs. Specifically, I was going to start splitting them into bins using make_taxonomy_table.py and I saw there is a way to direct to the coverage table with the  --cov_table flag. It would be ideal to just put the .tab file make for each sample after that flag but I'm not sure that will work without coverage values, so I wanted to see if there is an easy way to convert those lengths to coverage. Or rather if there is a way to get a coverage table from the 8 output files that were generated for each sample (6 bt2 files, 1 .bam, 1 .txt, 1 .tab). ?

**Reply by evanroyrees** [reply]
From your script, I'm guessing the coverage table was repeatedly overwritten. To help you out, you could try our most up-to-date code. Documentation for our updated code base may be found at [autometa.readthedocs.io](https://autometa.readthedocs.io)


Here is [the section for retrieving a formatted coverage table](https://autometa.readthedocs.io/en/latest/step-by-step-tutorial.html#coverage-calculation) at varying stages of coverage calculations. Since you have the bam files from each run, you can retrieve coverage tables from these. See [this section](https://autometa.readthedocs.io/en/latest/step-by-step-tutorial.html#from-alignments-bam) for how to do this.

This will require installation which may be found in the [installation docs section](https://autometa.readthedocs.io/en/latest/installation.html#install-from-source-using-make). Notice, you will need to install autometa from the git repository rather than the section labeled (quickest) as we have yet to release autometa on bioconda.

The section to which I'm referring is titled, "[Install from source (using make)](https://autometa.readthedocs.io/en/latest/installation.html#install-from-source-using-make)"

**Comment by Jmaccaro** [comment]
This is so helpful thank you so much!!

On Tue, Feb 1, 2022 at 3:32 PM Evan Rees ***@***.***> wrote:

> From your script, I'm guessing the coverage table was repeatedly
> overwritten. To help you out, you could try our most up-to-date code.
> Documentation for our updated code base may be found at
> autometa.readthedocs.io
>
> Here is the section for retrieving a formatted coverage table
> <https://autometa.readthedocs.io/en/latest/step-by-step-tutorial.html#coverage-calculation>
> at varying stages of coverage calculations.
>
> This will require installation which may be found in the installation
> docs section
> <https://autometa.readthedocs.io/en/latest/installation.html#install-from-source-using-make>.
> Notice, you will need to install autometa from the git repository rather
> than the section labeled (quickest) as we have yet to release autometa on
> bioconda.
>
> The section to which I'm referring is titled, "Install from source (using
> make)
> <https://autometa.readthedocs.io/en/latest/installation.html#install-from-source-using-make>
> "
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/KwanLab/Autometa/discussions/232#discussioncomment-2092589>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARIPOS6VZHKIVSRVYLJ5YPDUZBUQDANCNFSM5NJYLT5Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Jessica Maccaro
Ph.D. Student in McFrederick Lab
Department of Entomology
University of California, Riverside

**Comment by Jmaccaro** [comment]
Once I have downloaded the source code (I had to use the full commands option because I am running on a cluster and it didn't like the make option- they seemed the same since it was still putting the source code from the git repository)
I am still having problems with autometa-coverage 
I get this error: 
FileNotFoundError: [Errno 2] No such file or directory: '$PATHTOAUTOMETASOURCECODE/Autometa/cov-alignmen
ts97wwdqnl'

I ran this code in a similar loop to before. 
	autometa-coverage \
    	     --assembly $PATHTOASSEMBLIES/$assembly \
   	     --bam $PATHSTOBAMS/$bam \
             --out $PATHTOWHEREIWANTOUTPUT/Autometa/${samplename}.coverages.tsv
 
Here is the full output incase it helps: 
Traceback (most recent call last):
  File ".conda/envs/autometa2/bin/autometa-coverage", line 33, in <module>
    sys.exit(load_entry_point('Autometa==2.0.0', 'console_scripts', 'autometa-coverage')())
  File "$PATHTOAUTOMETASOURCECODE/Autometa/autometa/common/coverage.py", line 261, in main get(
  File "$PATHTOAUTOMETASOURCECODE/Autometa/autometa/common/utilities.py", line 422, in wrapper
    obj = func(*args, **kwds)
  File "$PATHTOAUTOMETASOURCECODE/Autometa/autometa/common/coverage.py", line 138, in get
    tempdir = tempfile.mkdtemp(suffix=None, prefix="cov-alignments", dir=outdir)
  File ".conda/envs/autometa2/lib/python3.9/tempfile.py", line 363, in mkdtemp
    _os.mkdir(file, 0o700)
FileNotFoundError: [Errno 2] No such file or directory: '$PATHTOAUTOMETASOURCECODE/Autometa/cov-alignments2huu5z6w'


Thanks again so much!

**Reply by evanroyrees** [reply]
Hmmm, I have a few questions.

1. Are there other logs emitted prior to the error? If so, could you share these? (If you are not comfortable sharing these in a public location, you may also email me at erees[at]wisc[dot]edu)

2. Does this directory exist `$PATHTOWHEREIWANTOUTPUT/Autometa`?

3. Just to double check, is the following a copy/paste mistake?

This is the last error you posted: 

```bash
FileNotFoundError: [Errno 2] No such file or directory: '$PATHTOAUTOMETASOURCECODE/Autometa/cov-alignments2huu5z6w'
```

However, I was expecting this error to emit:

```bash
FileNotFoundError: [Errno 2] No such file or directory: '$PATHTOWHEREIWANTOUTPUT/Autometa/cov-alignments2huu5z6w'
```

My first hunch is there is some sort of permissions error when attempting to create a temporary directory to perform calculations, but I want to check everything else is working appropriately before I go down that rabbit hole. 😅

**Comment by evanroyrees** [comment, marked as answer]
It turned out to be an incorrect file path ✅  on the end-user side. Whoops! 🙈
